File: 0015-3sum/0015-3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        
        int n=nums.size();
        vector<vector<int>>ans;
        set<vector<int>>set;
        sort(nums.begin(),nums.end());

        for(int i=0;i<n-2;i++)
        {
            int low=i+1, high=n-1;
            while(low<high)
            {
                if(nums[low]+nums[high]==(-1*nums[i]))
                {
                set.insert({nums[i],nums[low],nums[high]});
                low++;
                high--;
                }
                else if(nums[low]+nums[high]<(-1*nums[i]))
                {
                    low++;
                }
                else
                {
                    high--;
                }
            }
        }

        for(auto it: set)
        {
            ans.push_back(it);
        }

        return ans;
    }
};

File: 0015-3sum/0015-3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        
        Set<List<Integer>>result=new HashSet<>();
        if(nums.length<3)
        return new ArrayList<>(result);

        Arrays.sort(nums);
        for(int i=0;i<nums.length-2;i++){

            int sum=-1*nums[i], left=i+1, right=nums.length-1;
            while(left<right){
                int target=nums[left]+nums[right];
                if(sum==target){
                    result.add(Arrays.asList(nums[i],nums[left],nums[right]));
                    left++;
                    right--;
                }
                else if(sum<target){
                    right--;
                }
                else
                left++;
            }
        }

        return new ArrayList<>(result);
    }
}